assets/js/admin/admin.js

```diff
@@ -0,0 +1,117 @@
+        }
+
+        // Prepare the parameters for the AJAX Call.
+        return {
+            action: 'safe_svg_optimize',
+            svg_url: svgUrl,
+            optimized_svg: optimizedString,
+            svg_nonce: safeSvgParams.nonce
+        };
+    };
+
+    /**
+     * Hook into the WordPress Uploader and optimize the SVG.
+     */
+    Object.assign(
+        wp.Uploader.prototype, {
+            // Run on a successful upload.
+            success: function (attachment) {
+
+                const svgUrl = attachment?.attributes?.url;
+                if (!svgUrl || 'svg+xml' !== attachment?.attributes?.subtype) {
+                    return;
+                }
+
+                // Get the SVG data from the file's URL and optimize.
+                fetch(svgUrl, {method: 'GET'})
+                    .then((response) => response.text())
+                    .then((response) => {
+
+                        const params = ajaxUrlParams(svgUrl, response);
+                        if (!params) {
+                            return;
+                        }
+                        // Make an AJAX call to update the SVG file with the optimized contents.
+                        ajaxUrl.search = new URLSearchParams(params);
+                        fetch(ajaxUrl, {method: 'GET'})
+                            .then((ajaxResponse) => ajaxResponse)
+                            .then((ajaxResponse) => {
+                                if (200 !== ajaxResponse?.status) {
+                                    return;
+                                }
+
+                                // Refresh the uploader window to update the file size.
+                                if (wp.media.frame.content.get() !== null) {
+                                    wp.media.frame.content.get().collection.props.set({ignore: (+new Date())});
+                                    wp.media.frame.content.get().options.selection.reset();
+                                } else {
+                                    wp.media.frame.library.props.set({ignore: (+new Date())});
+                                }
+
+                            });
+                    });
+            }
+        }
+    );
+
+    /**
+     * Optimize the SVGs inserted in the editor.
+     * This takes care of SVGs uploaded via direct upload, without using the Media Library.
+     */
+    const {isSavingPost, getPostEdits} = select('core/editor');
+
+    subscribe(() => {
+        if (isSavingPost()) {
+            let changes = getPostEdits();
+            for (const changedBlock of changes.blocks) {
+
+                // Run only if this is a new core image block.
+                if ('core/image' === changedBlock.name && !changedBlock?.originalContent) {
+                    const svgUrl = changedBlock?.attributes?.url;
+
+                    // Run only if the image is an SVG that has not been optimized yet.
+                    if (!svgUrl || !svgUrl.endsWith('.svg')) {
```

Comment:
**thought (non-blocking):** This will only pick up files with the `.svg` extension. What about `.svgz` or SVGs that are uploaded without the SVG extension? 

Do we have access to the mime type at this point to check that instead?

Reply:
@darylldoyle We don't (that's the reason I made the check based on the filename). Here's an example of what the changedBlock returns:

![image](https://user-images.githubusercontent.com/5892287/198892775-ab866f1f-0a9c-4158-8e20-6fc43ed421f4.png)

One approach that I can think of is to remove that particular check and perform it inside the `fetch`, where the file's `content-type` becomes available. In that case, though, the `fetch` would have to run on every file, while now we only run it on .`svg` files. Perhaps we could mitigate that by narrowing it down to specific extensions (e.g. `svg*` for anything that ends in svg-something) and files with no extension at all. That way, we wouldn't bother running it on jpgs, pngs etc.

What do you think?